### ISSUE TITLE
Hybrid solvent model for protein simulation with openmm?

### ISSUE THREAD

**Issue by minhuanli** [openmm/openmm#3364]
Hello experts,

I've been trying to run simulations of proteins with a hybrid, both explicit+implicit, solvent model, but don't really know what is the right way to define the forcefield. 

It is common that most protein models have several explicit ordered waters listed as `HOH` residues in PDB files. For example, the [`7kr0`](https://www.rcsb.org/experimental/7kr0) protein. And besides the few ordered waters, the system has also been filled with random waters, which are usually modeled as implicit solvents.

However, for those models with explicit waters (already cleaned with [PDBfixer](https://github.com/openmm/pdbfixer)), directly using an implicit solvent forcefield will cause errors as following:
```
ff = ForceField('amber99sb.xml', 'amber99_obc.xml') # Implicit Solvent Model
pdb = PDBFile("7kr0_fixedwithHOH.pdb")
modeller = Modeller(pdb.topology, pdb.positions)
system = ff.createSystem(modeller.topology,
                         removeCMMotion=False,
                         nonbondedMethod=app.CutoffNonPeriodic, nonbondedCutoff=1.0*nanometers,
                         constraints=None, rigidWater=True)
```
Errors are like:
```
ValueError: No template found for residue 170 (HOH).  This might mean your input topology is missing some atoms or bonds, or possibly that you are using the wrong force field.
```
And if I added some explicit solvent template for the `HOH` residue, some other errors will happen:
```
ff = ForceField('amber99sb.xml', 'amber99_obc.xml', 'amber/tip3p_standard.xml') # Hybrid Solvent Model
pdb = PDBFile("7kr0_fixedwithHOH.pdb")
modeller = Modeller(pdb.topology, pdb.positions)
system = ff.createSystem(modeller.topology,
                         removeCMMotion=False,
                         nonbondedMethod=app.CutoffNonPeriodic, nonbondedCutoff=1.0*nanometers,
                         constraints=None, rigidWater=True)
```
Errors are like:
```
ValueError: GBSAOBCForce: No parameters defined for atom type tip3p-O
```

So I wonder if there is a way to do this hybrid solvent model simulation in openmm?

p.s.
The above mentioned protein model works with purely explicit solvent model, but it is not an accurate way as the system is also filled with random waters not listed in the PDB files.
```
ff = ForceField('amber99sb.xml', 'amber/tip3p_standard.xml') #Explicit model, template for HOH
pdb = PDBFile("7kr0_fixedwithHOH.pdb")
modeller = Modeller(pdb.topology, pdb.positions)
system = ff.createSystem(modeller.topology,
                         removeCMMotion=False,
                         nonbondedMethod=app.CutoffNonPeriodic, nonbondedCutoff=1.0*nanometers,
                         constraints=None, rigidWater=True)
```

**Comment by minhuanli** [comment]
Attach the example PDB files [here](https://drive.google.com/file/d/1xf_cb8Y-ONMbrCS_j2o6B5WHv8Qa2lFq/view?usp=sharing)

**Comment by peastman** [comment]
There are two questions here, one technical and one scientific.  I'll start with the technical one.

I think you should be able to do this with OpenMM 7.7.  We added support for implicit solvent with newer force fields, and the implementation works a bit differently.  I think the following ought to work.

```python
ff = ForceField('amber14-all.xml', 'amber14/tip3pfb.xml', 'implicit/obc2.xml')
```

There's a beta of 7.7 available if you want to try it.  The official release should be later this month.

Now the scientific question.  Hybrid explicit/implicit water is really hard to do accurately.  Some people have tried to develop methods for it, but I don't know of any that are widely used.  In your case you have a few structurally important waters that you want to model explicitly, since they make important specific interactions.  And you want to surround them with implicit solvent.  If you do that, the explicit waters will just float away.  There's nothing holding them in place.  Sometimes people try to use a droplet of explicit water, just enough to fully surround the solute, and apply some kind of restraint potential to keep them in place.  That can lead to artifacts both from the restraint potential and from the interface between explicit and implicit water.

Unless you're sure you know what you're doing, it's probably better to stick with just one water model, either explicit or implicit.

**Comment by lukeczapla** [comment]
On the scientific points, this sounds a bit similar to those spherical cells with the generalized solvent boundary potentials that have been floating around the literature forever, where your "droplet" is the cell and may not be spherical due to efficiency (depending on the shape of the protein).  It is totally doable by introducing potentials on the boundary and just thinking about this you might want to look into CustomExternalForce e.g. http://docs.openmm.org/latest/userguide/theory/03_custom_forces.html#customexternalforce

I think a lot of difficulties go into it to maintain properties such as pressure and things like that, and for that I would not have much to help but to suggest looking into the research papers with similar techniques like these spherical cells.   They never took off in popularity simply because the periodic boundary conditions with the PME was generally a lot quicker than electrostatics as an N^2 problem but physically it was interesting and worth looking at, especially with highly charged things like DNA where you may really want a lot of water in your periodic box, and with periodicity just a single unbalanced charge leads to the energy going off to infinity! (which is simple to derive from math)

**Comment by jchodera** [comment]
Here's a link to the [spherical solvent boundary potential (SSBP) method](https://doi.org/10.1063/1.466711) from Benoît Roux that @lukeczapla mentions in https://github.com/openmm/openmm/issues/3364#issuecomment-986171949, which was later generalized to non-spherical regions as the [generalized solvent boundary potential (GSBP)](https://doi.org/10.1063/1.1336570). These have indeed been around for quite a while, but have not seen much uptake, potentially due to complexities of implementation or use, but more likely due to the fundamental issue that @peastman points out: matching the interactions between explicit and implicit so that the boundary waters behave like bulk is really, really difficult.

The best new take on these ideas is to [simply use machine learning (ML) to solve this issue](https://arxiv.org/abs/2106.07492). OpenMM is already working to make it easy to integrate ML potentials at high speed via [openmm-torch](https://github.com/openmm/openmm-torch), so we expect there will be more development along these lines in the very near future.

**Comment by peastman** [comment]
Here are some papers on a more recent method, smoothly decoupled particle interface (SDPI).

https://aip.scitation.org/doi/abs/10.1063/1.3595262
https://aip.scitation.org/doi/abs/10.1063/1.4848655
https://aip.scitation.org/doi/full/10.1063/1.5025826

This method is a lot more complicated to implement, since it involves smoothly transitioning from explicit to implicit over a buffer region and continuously adding and removing water molecules.  At one point we were talking about trying to implement it in OpenMM, back when Jason Wagoner was in the Pande group, but nothing came of it.

**Comment by lukeczapla** [comment]
The methodology in the papers that @peastman shares above is also quite simply also called (or related to) water GCMC (grand canonical Monte Carlo) where I had a couple examples of inserting/deleting other things than water and setting the chemical potential correctly.  we had a paper with ion buffers on both sides of a nanopore where those buffer regions insert and remove ions to match the difference of concentrations across the pore, it's very similar to inserting waters but waters are more difficult and there it might be useful to look at the work of Mihaeli Mezei who had a package called MMC that at least shows the examples, which then are easily written with the OpenMM routines for acceleration.  

I wanted to use it in combination with drug docking to explicitly position the waters inside of the active site, because that's a major approximation there (SASA terms, cavitation energies, etc.), but it can be computationally expensive compared to some docking websites that actually did a fine job with even simpler versions of GBSA/MM like FACTS.  It'd be a fun project though because the waters are incredibly important in bridging certain H-bonds and such, and those imaginary terms like SASA and cavitation come from SOMEWHERE and are just being approximated by SASA and MV (molecular volume terms), I have to dig out a great review paper called something like "Tetrahedral condensed matter" about the whole water issue but it's super important!

**Comment by lukeczapla** [comment]
all of this is greatly implemented though in OpenMM, you might want the BrownianIntegrator to correctly model the dynamics http://docs.openmm.org/development/api-python/generated/openmm.openmm.BrownianIntegrator.html because the Langevin just has the sort of "random kick" term but not the solvent friction term, etc., a lot of important factors when different "droplets" are moving around and maybe then interacting, etc.  It depends on how complicated the model is.

**Comment by peastman** [comment]
Langevin does include friction.  You can set it to any value you want.  If you take the limit of very large friction, you get Brownian dynamics.

**Comment by lukeczapla** [comment]
yup that's correct, it's the overdamped langevin motion - forgot about that one 👍  but definitely the best setting for the dynamics here with these huge droplets

**Comment by peastman** [comment]
Brownian dynamics leads to much less efficient sampling than Langevin dynamics.  If you just care about sampling, not accurate dynamics, you definitely want Langevin.  And if you do care about accurate dynamics you still want Langevin, just with a larger friction coefficient.  The friction coefficient of bulk water is 91 ps^-1 which is large but nowhere near the overdamped limit at the atomic scale.

Of course, that refers to implicit solvent.  If you include explicit solvent, it already provides friction so you don't need as much external friction.

**Comment by lukeczapla** [comment]
I was just referring to if you want to move the entire droplet through space, the BrownianIntegrator would be great for modelling something like that.  two droplets far enough away in space with ionic conditions in the solution (Debye-Huckel) literally have no interaction so far away it's trivial and then as they get closer you could incorporate their interactions.

**Comment by peastman** [comment]
That's equally true for Langevin.

**Comment by lukeczapla** [comment]
it's true so in a sense you're saying it's ridiculous that BrownianIntegrator even exists

**Comment by peastman** [comment]
If someone really wants to simulate something in the overdamped limit, BrownianIntegrator is the right tool for it.  There are lots of physical systems that are well described by that limit.  Atomic scale motions in water aren't one of them.

**Comment by lukeczapla** [comment]
yep so for a giant protein droplet moving through space I was saying you could use the BrownianIntegrator to update everyone's position (you could say the center of mass of the thing as a giant particle) but then it seems you're also saying that you could tune LangevinIntegrator however you want to reproduce the same overdamped thing?  That's why it appears that it's a bit ridiculous that it even exists
